src/box/memtx_tx.c

```diff
@@ -2876,6 +2876,17 @@ memtx_tx_index_invisible_count_slow(struct txn *txn,
 		}
 		assert(link->newer_story == NULL);
 
+		/*
+		 * Excluded tuples have their own chains consisting of the only
+		 * excluded story. Such stories must be skipped since they are
+		 * not actually inserted to index.
+		 */
+		if (tuple_key_is_excluded(story->tuple, index->def->key_def,
```

Comment:
Ideally, these tuples should be filtered out by the check above (`link->in_index == NULL`), since they are not physically present in the index. See my comment above.

Reply:
I agree, but on the other hand we have an invariant that top story of every chain is physically in index, so lots of assertions check that `link->in_index != NULL` at the top story.

As a simplest solution, we could add a flag to `link` whether the story is excluded from the index, so we could replace assertions with `assert(link->in_index != NULL || link->is_excluded)`. But, probably, that's not the best solution and we can come up with something better.

Let's create another issue describing this problem and merge this fix as is. Are you okay with that?

Reply:
Yeah.

Reply:
OK, I'll file an issue.